NodeIPAM CloudAllocator Type on KCM panics when using external cloud provider

Cloud-providers started to move out of tree its controller, but they share some functionalities, one of them is the NodeIPAM controller.

There is a type that is GCE specific, the `CloudAllocator` type.

If this type is used, but no configuration for GCE is passed, the kcm panics.


This can be observed in the following job: https://testgrid.k8s.io/sig-testing-canaries#ingress-gce-e2e-canary

> I1217 13:13:43.823361       9 flags.go:64] FLAG: --cidr-allocator-type="CloudAllocator"

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x28a2621]

goroutine 205 [running]:
k8s.io/kubernetes/pkg/controller/nodeipam/ipam.NewCloudCIDRAllocator({0x51dd7f0, 0xc000f2c000}, {0x0, 0x0?}, {0x5194a40, 0xc000dffd40})
	pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go:98 +0x5e1
k8s.io/kubernetes/pkg/controller/nodeipam/ipam.New({0x51dd7f0, 0xc000f2c000}, {0x0, 0x0}, {0x5194a40, 0xc000dffd40}, {0x0, 0x0}, {0x7fffae6b9d76, 0xe}, ...)
	pkg/controller/nodeipam/ipam/cidr_allocator.go:138 +0xed
k8s.io/kubernetes/pkg/controller/nodeipam.NewNodeIpamController({0x5194a40, 0xc000dffd40}, {0x0, 0x0}, {0x0?, 0x0}, {0x51dd7f0, 0xc000f2c000}, {0xc000700ea0, 0x1, ...}, ...)
	pkg/controller/nodeipam/node_ipam_controller.go:134 +0x665
k8s.io/kubernetes/cmd/kube-controller-manager/app.startNodeIpamController({_, _}, {{0x51b8a40, 0xc000087e60}, {0x51d33d0, 0xc00016ac40}, {0x5196e30, 0xc000c23120}, {{{0x0, 0x0}, ...}, ...}, ...})
	cmd/kube-controller-manager/app/core.go:152 +0x666
k8s.io/kubernetes/cmd/kube-controller-manager/app.StartControllers({_, _}, {{0x51b8a40, 0xc000087e60}, {0x51d33d0, 0xc00016ac40}, {0x5196e30, 0xc000c23120}, {{{0x0, 0x0}, ...}, ...}, ...}, ...)
	cmd/kube-controller-manager/app/controllermanager.go:594 +0x4e4
k8s.io/kubernetes/cmd/kube-controller-manager/app.Run.func1({0x51b04b0, 0xc0005d4a80}, 0xc00031a570?, 0x4b71a98)
	cmd/kube-controller-manager/app/controllermanager.go:231 +0x26e
k8s.io/kubernetes/cmd/kube-controller-manager/app.Run.func3({0x51b04b0?, 0xc0005d4a80?})
	cmd/kube-controller-manager/app/controllermanager.go:291 +0xe3
created by k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run
	vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:211 +0x11b
```

The question is, should this type be removed or a noop in KCM?

--- a/cmd/cloud-controller-manager/nodeipamcontroller.go
+++ b/cmd/cloud-controller-manager/nodeipamcontroller.go
@@ -75,6 +75,11 @@ func startNodeIpamController(initContext app.ControllerInitContext, ccmConfig *c
 		return nil, false, nil
 	}
 
+	// Cannot run cloud ipam controller if cloud provider is nil (--cloud-provider not set or set to 'external')
+	if cloud == nil && ccmConfig.ComponentConfig.KubeCloudShared.CIDRAllocatorType == string(ipam.CloudAllocatorType) {
+		return nil, false, errors.New("--cidr-allocator-type is set to 'CloudAllocator' but cloud provider is not configured")
+	}
+
 	// failure: bad cidrs in config
 	clusterCIDRs, dualStack, err := processCIDRs(ccmConfig.ComponentConfig.KubeCloudShared.ClusterCIDR)
 	if err != nil {

--- a/cmd/kube-controller-manager/app/core.go
+++ b/cmd/kube-controller-manager/app/core.go
@@ -105,6 +105,11 @@ func startNodeIpamController(ctx context.Context, controllerContext ControllerCo
 		return nil, false, nil
 	}
 
+	// Cannot run cloud ipam controller if cloud provider is nil (--cloud-provider not set or set to 'external')
+	if controllerContext.Cloud == nil && controllerContext.ComponentConfig.KubeCloudShared.CIDRAllocatorType == string(ipam.CloudAllocatorType) {
+		return nil, false, errors.New("--cidr-allocator-type is set to 'CloudAllocator' but cloud provider is not configured")
+	}
+
 	clusterCIDRs, err := validateCIDRs(controllerContext.ComponentConfig.KubeCloudShared.ClusterCIDR)
 	if err != nil {
 		return nil, false, err

